xarray/conventions.py

```diff
@@ -139,6 +139,15 @@ def _infer_dtype(array, name=None):
     if array.size == 0:
         return np.dtype(float)
 
+    native_dtypes = set(np.vectorize(type, otypes=[object])(array.ravel()))
+    if len(native_dtypes) > 1 and native_dtypes != {bytes, str}:
+        raise ValueError(
+            "unable to infer dtype on variable {!r}; object array "
+            "contains mixed native types: {}".format(
+                name, ", ".join(map(lambda x: x.__name__, native_dtypes))
```

Comment:
Nit, feel free to ignore

```suggestion
                name, ", ".join(x.__name__ for x in native_dtypes))
```